datafusion/physical-plan/src/joins/dynamic_filters.rs

```diff
@@ -0,0 +1,267 @@
+                    Ok((max_acc, min_acc))
+                },
+            )?;
+
+        Ok(Self {
+            columns,
+            build_side_names,
+            inner: Mutex::new(DynamicFilterInfoInner {
+                max_accumulators,
+                min_accumulators,
+                final_expr: None,
+                batch_count: total_batches,
+                processed_partitions: HashSet::with_capacity(total_batches),
+            }),
+        })
+    }
+
+    // Set the final expression for the filter
+    pub fn with_final_expr(self, expr: Arc<dyn PhysicalExpr>) -> Self {
+        self.inner.lock().final_expr = Some(expr);
+        self
+    }
+
+    // Merge a new batch of data and update the max and min accumulators
+    pub fn merge_batch_and_check_finalized(
+        &self,
+        records: &RecordBatch,
+        partition: usize,
+    ) -> Result<bool, DataFusionError> {
+        let mut inner = self.inner.lock();
+
+        if inner.final_expr.is_some() {
+            return Ok(true);
+        }
+
+        if !inner.processed_partitions.insert(partition) {
+            return Ok(false);
+        }
+
+        inner.batch_count = inner.batch_count.saturating_sub(1);
+        if records.num_rows() == 0 {
+            return Ok(false);
+        }
+
+        let finalize = inner.batch_count == 0;
+
+        let schema = records.schema();
+        let columns = records.columns();
+        for (i, _) in self.columns.iter().enumerate() {
+            let index = schema.index_of(&self.build_side_names[i])?;
+            let column_data = &columns[index];
+            inner.max_accumulators[i]
+                .update_batch(&[Arc::<dyn arrow_array::Array>::clone(column_data)])?;
+            inner.min_accumulators[i]
+                .update_batch(&[Arc::<dyn arrow_array::Array>::clone(column_data)])?;
+        }
+
+        if finalize {
+            drop(inner);
+            self.finalize_filter()?;
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+
+    // Finalize the filter by creating the final expression
+    fn finalize_filter(&self) -> Result<(), DataFusionError> {
+        let mut inner = self.inner.lock();
+        let filter_expr =
+            self.columns.iter().enumerate().try_fold::<_, _, Result<
+                Option<Arc<dyn PhysicalExpr>>,
+                DataFusionError,
+            >>(None, |acc, (i, column)| {
+                let max_value = inner.max_accumulators[i].evaluate()?;
+                let min_value = inner.min_accumulators[i].evaluate()?;
+
+                let max_scalar = max_value.clone();
+                let min_scalar = min_value.clone();
+
+                let max_expr: Arc<dyn PhysicalExpr> = Arc::new(Literal::new(max_scalar));
+                let min_expr: Arc<dyn PhysicalExpr> = Arc::new(Literal::new(min_scalar));
+
+                let range_condition: Arc<dyn PhysicalExpr> = Arc::new(BinaryExpr::new(
```

Comment:
Maybe it makes sense to not create a dynamic filter in certain conditions, eg `min==max` (will not filter out any rows)?